pkg/spanconfig/spanconfigsqlwatcher/sql_watcher.go

```diff
@@ -0,0 +1,453 @@
+
+		var id descpb.ID
+		if table != nil {
+			id = table.GetID()
+		} else if database != nil {
+			id = database.GetID()
+		} else {
+			// We only care about database or table descriptors being updated.
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+		case updatesCh <- id:
+		}
+	}
+	opts := []rangefeed.Option{
+		rangefeed.WithDiff(),
+	}
+	if !s.knobs.SQLWatcherDisableInitialScan {
+		opts = append(opts, rangefeed.WithInitialScan(nil))
+		opts = append(opts, rangefeed.WithOnInitialScanError(func(ctx context.Context, err error) bool {
+			logcrash.ReportOrPanic(
+				ctx,
+				&s.settings.SV,
+				"error performing full reconciliation: %v",
+				err,
+			)
+			return false
+		}))
+	}
+	rf, err := s.rangeFeedFactory.RangeFeed(
+		ctx,
+		"sql-watcher-descriptor-rangefeed",
+		descriptorTableSpan,
+		s.clock.Now(),
+		handleEvent,
+		opts...,
+	)
+	if err != nil {
+		return nil, err
+	}
+	s.stopper.AddCloser(rf)
+
+	log.Infof(ctx, "established range feed over system.descriptors table")
+	return updatesCh, nil
+}
+
+// watchForZoneConfigUpdates establishes a rangefeed over system.zones and sends
+// updates on the returned channel.
+func (s *SQLWatcher) watchForZoneConfigUpdates(ctx context.Context) (<-chan descpb.ID, error) {
+	updatesCh := make(chan descpb.ID)
+
+	zoneTableStart := s.codec.TablePrefix(keys.ZonesTableID)
+	descriptorTableSpan := roachpb.Span{
+		Key:    zoneTableStart,
+		EndKey: zoneTableStart.PrefixEnd(),
+	}
+
+	handleEvent := func(ctx context.Context, ev *roachpb.RangeFeedValue) {
+		var descID descpb.ID
+		{
+			// Decode the descriptor ID from the key.
+			tbl := systemschema.ZonesTable
+			types := []*types.T{tbl.PublicColumns()[0].GetType()}
+			startKeyRow := make([]rowenc.EncDatum, 1)
+			_, matches, _, err := rowenc.DecodeIndexKey(
+				s.codec, tbl, tbl.GetPrimaryIndex(),
+				types, startKeyRow, nil, ev.Key,
+			)
+			if err != nil || !matches {
+				logcrash.ReportOrPanic(
+					ctx,
+					&s.settings.SV,
+					"failed to decode key in system.zones: %v",
+					ev.Key,
+				)
+				return
+			}
+			alloc := &rowenc.DatumAlloc{}
+			if err := startKeyRow[0].EnsureDecoded(types[0], alloc); err != nil {
+				logcrash.ReportOrPanic(
+					ctx,
+					&s.settings.SV,
+					"failed to decode key in system.zones: %v",
+					ev.Key,
+				)
+				return
+			}
+			descID = descpb.ID(tree.MustBeDInt(startKeyRow[0].Datum))
+		}
+
+		select {
+		case <-ctx.Done():
+		case updatesCh <- descID:
+		}
+	}
+	rf, err := s.rangeFeedFactory.RangeFeed(
+		ctx,
+		"sql-watcher-zones-rangefeed",
+		descriptorTableSpan,
+		s.clock.Now(),
+		handleEvent,
+	)
+	if err != nil {
+		return nil, err
+	}
+	s.stopper.AddCloser(rf)
+
+	log.Infof(ctx, "established range feed over system.zones table")
+	return updatesCh, nil
+}
+
+// getAllAffectedDescriptors returns a list of IDs that need to have their span
+// configurations refreshed because the given id changed.
+func getAllAffectedIDs(
```

Comment:
The IDs returned are all table IDs, name the function accordingly. getAffectedTableIDs? Move this below onDescIDUpdate, where this is called.